falcon-taz/src/main/java/pt/haslab/taz/TraceProcessor.java

```diff
@@ -240,17 +241,20 @@ private void parseJSONEvent( JSONObject event )
         String thread = event.getString( "thread" );
         String loc = event.optString( "loc" );
         loc = ( loc == null ) ? "" : loc;
+        // consider timestamp to be a long for the moment
         long time = event.getLong( "timestamp" );
         String timestamp = String.valueOf( time );
-        //** consider timestamp to be a long for the moment
-        //BigDecimal bd = new BigDecimal(String.valueOf(time)).multiply(new BigDecimal(1000000)).stripTrailingZeros();
-        //String timestamp = bd.toPlainString();
-        eventNumber++;
 
-        Event e = new Event( timestamp, type, thread, eventNumber, loc );
+        // --- OPTIONAL FIELDS ---
+        //use event id from trace if it's present in the JSON object
+        //otherwise, use global event counter as id
+        long eventId = event.has( "id" ) ? event.getLong( "id" ) : eventNumber++;
+        Event e = new Event( timestamp, type, thread, eventId, loc );
 
-        //optional fields
-        e.setDependency( event.optString( "dependency", null ) );
+        String dependency = event.optString( "dependency" );
+        if(dependency.length() == 0 || dependency.equals( "null" ))
```

Comment:
Why is it better than: `event.optString( "dependency", null )` ?

I mean, having the string `"null"` in the dependency field should not happen at all.

Reply:
falcon-solver outputs JSON objects with `"dependency":"null"` if it receives as input a trace file with `"dependency":null`. This code fixes that.

Reply:
Maybe we should try `getString` with the default value as `null`. Does `getString` throw an exception if it is explicitly `null` ?

Reply:
`optString( "dependency" )` returns `"null"` if the field is `"dependency":null`